src/site_apportionment/site_apportionment.py

```diff
@@ -0,0 +1,222 @@
+
+    # df_cols: original columns
+    df_cols = list(df.columns)
+
+    # Create a list of the value columns that we want to apportion
+    # These are the same as the columns we impute so we use a function from imputation.
+    value_cols = get_imputation_cols(config)
+
+    # Calculate the number of unique non-blank postcodes
+    df = count_unique_codes_in_col(df, postcode)
+
+    # Condition of long forms, many sites, instance >=1, non-null postcodes
+    cond = (
+        (df[form] == long_code) &
+        (df[postcode + "_count"] > 1) &
+        (df[ins] >= 1)
+    )
+
+    # Dataframe dfm with many products - for apportionment and Cartesian product
+    dfm = df.copy()[cond]
+    dfm = dfm.drop(columns=[postcode + "_count"], axis=1)
+
+    # Dataframe with everything else - save unchanged
+    df_out = df[~cond]
+    df_out = df_out.drop(columns=[postcode + "_count"], axis=1)
+
+    # df_codes: dataframe with codes and numerical values
+    group_cols = [ref, period]
+    code_cols = [product, civdef, pg_num]
+    df_codes = dfm.copy()[group_cols + code_cols + value_cols]
+
+    # Remove blank products
+    df_codes = df_codes[df_codes[product].str.len() > 0]
+
+    # # De-duplicate by summation - possibly, not needed
+    df_codes = (
+        df_codes.groupby(group_cols + code_cols).agg(sum).reset_index()
+    )
+
+    # df_sites: dataframe with postcodes, percents, and everyting else
+    site_cols = [x for x in df_cols if x not in (code_cols + value_cols)]
+    df_sites = dfm.copy()[site_cols]
+
+    # Remove instances that have no postcodes
+    df_sites = df_sites[df_sites[postcode].str.len() > 0]
+
+    # Check for postcode duplicates for QA
+    df_sites["site_count"] = (
+        df_sites.groupby(group_cols + [postcode])[postcode].transform("count")
+    )
+    df_duplicate_sites = df_sites[df_sites["site_count"] > 1]
+    num_duplicate_sites = df_duplicate_sites.shape[0]
+    if num_duplicate_sites:
+        SitesApportionmentLogger.info(
+            f"There are {num_duplicate_sites} duplicate sites."
+        )
+
+    # Calculate weights
+    df_sites = weights(df_sites)
+
+    #  Merge codes to sites to create a Cartesian product
+    df_cart = df_sites.merge(df_codes, on=group_cols, how="inner")
+
+    # Apply weights
+    for value_col in value_cols:
+        df_cart[value_col] = df_cart[value_col] * df_cart["site_weight"]
+
+    # Restore the original column order
+    df_cart = df_cart[df_cols]
+
+    # Append the apportionned data back to the remaining unchanged data
+    df_out = df_out.append(df_cart, ignore_index=True)
+
+    # Sort by period, ref, instance in ascending order.
+    df_out.sort_values(by=[period, ref, ins], ascending=True, inplace=True)
+
+    return df_out
+
+
+if __name__ == "__main__":
+    mydir = r"D:\data\res_dev\outputs\reg_apport"
+    in_file = "outputs_df_before.pkl"
+    out_file = "df_out.csv"
+    mypath = os.path.join(mydir, in_file)
+    df = pd.read_pickle(mypath)
```

Comment:
Why are we using pickle here and feather everywhere else? I think it's good to be consistent.

Reply:
Removed this whole block, as it was only needed for initial development and debugging

Reply:
This was for running the script on its own, not as part of the pipeline. Removed now